Espressif ESP32 Benchmark, Test, TLS 1.3 Client & Server Updates

gojimmypi opened the pull request:
# Description

This PR updates the Espressif ESP32 TLS [Client](https://github.com/wolfSSL/wolfssl/tree/master/IDE/Espressif/ESP-IDF/examples/wolfssl_client) and [Server](https://github.com/wolfSSL/wolfssl/tree/master/IDE/Espressif/ESP-IDF/examples/wolfssl_server) example apps.

Of particular interest is the new `CMakeLists.txt` that _searches_ for wolfSSL source code and expects the `user_settings.h` to now be in the project directory: `components/wolfssl/include`. See https://github.com/wolfSSL/wolfssl/issues/6118.

The update also adds support for the ESP-IDF v5 (specifically v5.1) as noted in  https://github.com/wolfSSL/wolfssl/issues/5909.

The Wi-Fi networking connection has been improved, using example ESP-IDF code when possible. There's also improved time-setting and NTP support, falling back to GitHub recent commit date, and then worst case to a fixed date time value.

 SM Cipher support has been added. See https://github.com/wolfSSL/wolfssl/pull/6671 and [wolfSSL/wolfsm](https://github.com/wolfSSL/wolfsm). The SM ciphers need to be manually installed as noted in the [README.md](https://github.com/wolfSSL/wolfsm#install-sm-code-into-wolfssl). Enable SM in the local `user_settings.h` file. The [wolfssl/certs_test_sm.h](https://github.com/gojimmypi/wolfssl/blob/ED25519_SHA2_fix/wolfssl/certs_test_sm.h) is also needed from https://github.com/wolfSSL/wolfssl/issues/6681 / https://github.com/wolfSSL/wolfssl/pull/6825 for these examples to work properly with SM ciphers.

* Edit: This PR now includes changes to support ESP-IDF v4.4, 5.0, and 5.1. See [Espressif supported version list](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/versions.html).

* Edit (2): I've created https://github.com/wolfSSL/wolfssl/pull/6877 to fix the Jenkins build failures in this PR.

* Edit(3): This PR now updates all Espressif examples in preparation for  https://github.com/wolfSSL/wolfssl/pull/6990

Fixes zd#  n/a

See https://github.com/wolfSSL/wolfssl/issues/6234 for a roadmap of Espressif updates.


# Testing

How did you test?

Manually testing in VisualGDB and using ESP-IDF from command prompt.

# Checklist

 - [ ] added tests
 - [ ] updated/added doxygen
 - [ ] updated appropriate READMEs
 - [ ] Updated manual and documentation

gojimmypi commented:
Jenkins retest this please

gojimmypi commented:
Jenkins retest this please

JacobBarthelmeh commented on IDE/Espressif/ESP-IDF/examples/wolfssl_server/main/Kconfig.projbuild:
```diff
@@ -0,0 +1,29 @@
+menu "Example Configuration"
```
Is this right for in the wolfssl_server directory? Looks like benchmark options.

gojimmypi replied:
Good catch! That's correct - it was a copy of `Kconfig.projbuild` with the intention of having all the examples have kconfig capabilities. I've fixed that, along with other polishing updates to the examples. Once I've confirmed all is well, I'll push an additional commit.

JacobBarthelmeh commented on IDE/Espressif/ESP-IDF/examples/wolfssl_client/main/time_helper.c:
```diff
@@ -0,0 +1,333 @@
+/* time_helper.c
```
This (time_helper.c) and wifi_connect.c look to be duplicated with wolfssl_client and wolfssl_server. Not critical, and could be approved with out changing, but please take a look at combining them for ease of maintenance. Having the client and server use the same time_helper.c and wifi_connect.c files.

gojimmypi replied:
Hi @JacobBarthelmeh, you are correct: the code is duplicated between both the client and server applications. It was always duplicated in the `main.c`, but now it is more obvious in separate files.

I prefer to have the examples ready-to-run, fresh out of the source code, with no additional steps for install/setup (well, other than setting up the external ESP-IDF environment). These are also examples that are published as-is to the Espressif component registry. So it seems best to have the duplicate source code in order to have each example be stand-alone.

I do agree with the duplication and maintenance annoyance. @bandi13 also pointed that out in a recent PR of mine. 

Other than moving it to something like `esp_util.c` in wolfCrypt (where it arguably may not belong), do you have a suggestion where "_common-but-not-wolfssl_" code like this should live? Perhaps just move both the client and server to a shared directory with two separate projects but otherwise common code? This would be different than all the other examples. Completely open to suggestions here.

gojimmypi commented:
The latest commit addresses code review, polishes examples, and updates for these SoC devices:

- esp32
- esp32s2
- esp32c3
- esp32s3
- esp32c2
- esp32c6
- esp32h2

 For ESP32-C2 support, a future separate PR will be needed to actually enable HW encryption. (coming soon)
